lib/src/aarch64/gainmapmath_neon.cpp

```diff
@@ -0,0 +1,242 @@
+  return v_output;
+}
+
+int16x8x3_t yuvConversion_neon(uint8x8_t y, int16x8_t u, int16x8_t v, int16x8_t coeffs) {
+  const int16x8_t y_output = yConversion_neon(y, u, v, coeffs);
+  const int16x8_t u_output = uConversion_neon(u, v, coeffs);
+  const int16x8_t v_output = vConversion_neon(u, v, coeffs);
+  return {y_output, u_output, v_output};
+}
+
+void transformYuv420_neon(jr_uncompressed_ptr image, const int16_t* coeffs_ptr) {
+  // Implementation assumes image buffer is multiple of 16.
+  assert(image->width % 16 == 0);
+  uint8_t* y0_ptr = static_cast<uint8_t*>(image->data);
+  uint8_t* y1_ptr = y0_ptr + image->luma_stride;
+  uint8_t* u_ptr = static_cast<uint8_t*>(image->chroma_data);
+  uint8_t* v_ptr = u_ptr + image->chroma_stride * (image->height / 2);
+
+  const int16x8_t coeffs = vld1q_s16(coeffs_ptr);
+  const uint16x8_t uv_bias = vreinterpretq_u16_s16(vdupq_n_s16(-128));
+  size_t h = 0;
+  do {
+    size_t w = 0;
+    do {
+      uint8x16_t y0 = vld1q_u8(y0_ptr + w * 2);
+      uint8x16_t y1 = vld1q_u8(y1_ptr + w * 2);
+      uint8x8_t u = vld1_u8(u_ptr + w);
+      uint8x8_t v = vld1_u8(v_ptr + w);
+
+      // 128 bias for UV given we are using libjpeg; see:
+      // https://github.com/kornelski/libjpeg/blob/master/structure.doc
+      int16x8_t u_wide_s16 = vreinterpretq_s16_u16(vaddw_u8(uv_bias, u));  // -128 + u
+      int16x8_t v_wide_s16 = vreinterpretq_s16_u16(vaddw_u8(uv_bias, v));  // -128 + v
+
+      const int16x8_t u_wide_lo = vzip1q_s16(u_wide_s16, u_wide_s16);
+      const int16x8_t u_wide_hi = vzip2q_s16(u_wide_s16, u_wide_s16);
+      const int16x8_t v_wide_lo = vzip1q_s16(v_wide_s16, v_wide_s16);
+      const int16x8_t v_wide_hi = vzip2q_s16(v_wide_s16, v_wide_s16);
+
+      const int16x8_t y0_lo = yConversion_neon(vget_low_u8(y0), u_wide_lo, v_wide_lo, coeffs);
+      const int16x8_t y0_hi = yConversion_neon(vget_high_u8(y0), u_wide_hi, v_wide_hi, coeffs);
+      const int16x8_t y1_lo = yConversion_neon(vget_low_u8(y1), u_wide_lo, v_wide_lo, coeffs);
+      const int16x8_t y1_hi = yConversion_neon(vget_high_u8(y1), u_wide_hi, v_wide_hi, coeffs);
+
+      const int16x8_t new_u = uConversion_neon(u_wide_s16, v_wide_s16, coeffs);
+      const int16x8_t new_v = vConversion_neon(u_wide_s16, v_wide_s16, coeffs);
+
+      // Narrow from 16-bit to 8-bit with saturation.
+      const uint8x16_t y0_output = vcombine_u8(vqmovun_s16(y0_lo), vqmovun_s16(y0_hi));
+      const uint8x16_t y1_output = vcombine_u8(vqmovun_s16(y1_lo), vqmovun_s16(y1_hi));
+      const uint8x8_t u_output = vqmovun_s16(vaddq_s16(new_u, vdupq_n_s16(128)));
+      const uint8x8_t v_output = vqmovun_s16(vaddq_s16(new_v, vdupq_n_s16(128)));
+
+      vst1q_u8(y0_ptr + w * 2, y0_output);
+      vst1q_u8(y1_ptr + w * 2, y1_output);
+      vst1_u8(u_ptr + w, u_output);
+      vst1_u8(v_ptr + w, v_output);
+
+      w += 8;
+    } while (w < image->width / 2);
+    y0_ptr += image->luma_stride * 2;
+    y1_ptr += image->luma_stride * 2;
```

Comment:
I think this may go out of boundary if the image height is odd. Could you double check?

Reply:
This change conflicts with https://github.com/google/libultrahdr/pull/126/commits
@cmacdonald-arm, few observations, 
- folder structure is slightly different. I am using lib/src/dsp/.  Is lib/src/dsp/arm/  better ? (will be similar to libvpx)
- enabled aosp build (Android.bp as well)
- option to disable all intrinsics via macro at configure level
- support all resolutions, widths may not align to 16 always, example 1080x1920.
pls let me know if you have any suggestions.

Reply:
Hi Ram,

Personally I have no issues with the creation of a dsp/arm directory. I named the directory AArch64 to signify that all code is only 64-bit compatible. Also if some redesign is taking place to help with the integration of Neon implementations this could help greatly.

With your final point, images which are not aligned to 16 should be fine as along as the buffer containing the images is padded? Currently the ALIGNM macro is used to allow a multiple of 16 scan-lines to be passed to the jpeg encoder, is this likely to change?

ALIGNM is used in API-0: https://github.com/google/libultrahdr/blob/main/lib/src/jpegr.cpp#L225 
and API-1: https://github.com/google/libultrahdr/blob/main/lib/src/jpegr.cpp#L225

Reply:
aligned to 16 is an acceptable assumption. This is not likely to change. So the intrinsics should be fine. I have updated the folder structure to lib/src/dsp/arm/*.cpp

Reply:
Hi Ram,

Would the prefered approach here be, first merge your [PR](https://github.com/google/libultrahdr/pull/126/commits), with the updates to the build system and file structure/names. Second rebase these changes and rename the aarch64 Neon files?

Reply:
As pull request #126 contains build changes for different platforms (android and host), may be, it can be merged first. Then this change can be rebased on top of main.
We have introduced a macro UHDR_ENABLE_INTRINSICS to enable/disable intrinsics if required. I think that also needs to be incorporated in your change. Currently the __arch64__ is the only gating option.
There is gainmapmath_neon.h, can this be merged with gainmapmath.h just like editor_helper.h. This is to avoid more header files. If not, when x86 intrinsics are introduced there will be more header files.

Reply:
The change looks good to me. Thank you.